python/llm/README.md

```diff
@@ -1,47 +1,129 @@
 
 ```bash
-llm-cli.sh -x llama -h
+# text completion
+llm-cli -t 16 -x llama -m "/path/to/llama-7b-int4/bigdl-llm-xxx.bin" -p 'Once upon a time,'
+
+# chatting
+llm-cli -t 16 -x llama -m "/path/to/llama-7b-int4/bigdl-llm-xxx.bin" -i --color
 
-# Output:
-# usage: main-llama [options]
-#
-# options:
-#   -h, --help show this help message and exit
-#   -i, --interactive run in interactive mode
-#   --interactive-first run in interactive mode and wait for input right away
-#   ...
+# help information
+llm-cli -x llama -h
 ```
 
-### Examples
+#### Transformers like API
+Users could load converted model or even the unconverted huggingface model directly by `AutoModelForCausalLM.from_pretrained`.
 
-Here are some examples of how to use the llm-cli tool:
+```python
+from bigdl.llm.ggml.transformers import AutoModelForCausalLM
 
-#### Completion:
+# option 1: load converted model
+llm = AutoModelForCausalLM.from_pretrained("/path/to/llama-7b-int4/bigdl-llm-xxx.bin",
+                                           model_family="llama")
 
-```bash
-llm-cli.sh -t 16 -x llama -m ./llm-llama-model.bin -p 'Once upon a time,'
+# option 2: load huggingface checkpoint
+llm = AutoModelForCausalLM.from_pretrained("/path/to/llama-7b-hf/",
+                                           model_family="llama")
+
+# option 3: load from huggingface hub repo
+llm = AutoModelForCausalLM.from_pretrained("decapoda-research/llama-7b-hf",
+                                           model_family="llama")
 ```
 
-#### Chatting:
+Users could use llm to do the inference. Apart from end-to-end fast forward, we also support split the tokenization and model inference in our API.
 
-```bash
-llm-cli.sh -t 16 -x llama -m ./llm-llama-model.bin -i --color
+```python
+# end-to-end fast forward w/o spliting the tokenization and model inferencing
+result = llm("what is ai")
+
+# Use transformers tokenizer
+tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
+tokens = tokenizer("what is ai").input_ids
+tokens_id = llm.generate(tokens, max_new_tokens=32)
+tokenizer.batch_decode(tokens_id)
+
+# Use bigdl-llm tokenizer
+tokens = llm.tokenize("what is ai")
+tokens_id = llm.generate(tokens, max_new_tokens=32)
+decoded = llm.batch_decode(tokens_id)
+```
+
+#### llama-cpp-python like API
+`llama-cpp-python` has become a popular pybinding for `llama.cpp` program. Some users may be familiar with this API so `bigdl-llm` reserve this API and extend it to other model families (e.g., gptneox, bloom)
+
+```python
+from bigdl.llm.models import Llama, Bloom, Gptneox
+
+llm = Llama("/path/to/llama-7b-int4/bigdl-llm-xxx.bin", n_threads=4)
+result = llm("what is ai")
 ```
 
-Feel free to explore different options and experiment with the llama/gptneox/bloom models using
-llm-cli!
\ No newline at end of file
+#### langchain integration
+TODO
+
+## Examples
+We prepared several examples in https://github.com/intel-analytics/BigDL/tree/main/python/llm/example
+
+## Dynamic library BOM
```

Comment:
Is this dependency on user machine?

Reply:
This dependency is not needed for users. But I provide these information for compatibility reason. Users who are using lower version `GLIBC` may meet compatibility issue.

The version we use to build the dynamic lib and executable files are "as-low-as-possible" to increase the compatibility.

Reply:
> This dependency is not needed for users. But I provide these information for compatibility reason. Users who are using lower version `GLIBC` may meet compatibility issue.
> 
> The version we use to build the dynamic lib and executable files are "as-low-as-possible" to increase the compatibility.

SO this is actually user system dependency - for linux users they need to install glibc v9+, yes? how about windows users?

Reply:
Yeah, I will add this requirement. The upgrade of GLIBC could be really tricky so most of our users will use the default GLIBC version provided by their OS. In this case, I will state we support Windows 10/11 (10 not validated), Ubuntu 20.04 or later, CentOS 7 or later (bloom could not support CentOS).